### PR TITLE
[RSS] Require an explicit set of Bootstrap Agents

### PR DESCRIPTION
**Pull request by smklein**
Requires the client of RSS to supply provide the set of Bootstrap Agents which should be used for sled management and service provisioning.

This can be one of the two options

- `OnlyOurs` -- Ignore all other sleds but the one running RSS
- `OnlyThese` -- The caller provides the exact set of bootstrap addresses to be used.

I expected that the `OnlyOurs` option will be common for development, but the `OnlyThese` option will be more common when interacting with wicketd. Notably, this will require wicketd to explicitly select the set of sleds which should be used by RSS, referencing them by bootstrap address.

Fixes https://github.com/oxidecomputer/omicron/issues/2952

### CODE REVIEW

**Comment by jmpesp** [comment]
Could this be optional instead?

**Comment by smklein** [comment]
> Could this be optional instead?

The two options I provided are "pick your own", or "select exactly the set of bootstrap addresses you want".

Are you suggesting adding the old behavior of "include as many as you can see?"

**Comment by smklein** [comment]
> Sorry, I should have been more clear - I am, yeah. It seems like explicitly picking sleds like this should be temporary, until trust quorum / remote attestation comes along, so I thought it would be a good idea to keep the existing functionality.

I'm not certain this will be temporary. If wicketd is responsible for presenting a UI to an operator, and letting them update / select sleds which should be initialized, there's no reason why this couldn't be long-lived.

I think trust quorum / remote attestation could be layered on top of this -- we can still validate the sleds which the operator has explicitly selected -- but I don't think these options are mutually exclusive.

**Comment by rmustacc** [comment]
Right, I agree with @smklein this is not temporary. We always need to know what sleds we should be setting up with. Trust quorum doesn't change this, rather this is an input to trust quorum. If we're in a cold start case post-setup then I suspect things will be different in general as we're not going to be setting things up again.

**Comment by andrewjstone** [comment]
> Right, I agree with @smklein this is not temporary. We always need to know what sleds we should be setting up with. Trust quorum doesn't change this, rather this is an input to trust quorum. If we're in a cold start case post-setup then I suspect things will be different in general as we're not going to be setting things up again.

I'll add a third vote for this path forward. The way trust quorum works is that we need to know which sleds exist and should be part of the trust quorum, so that the correct number of key shares can be created and distributed. Discovery and sled-selection by an operator always comes first.

File: sled-agent/src/rack_setup/service.rs [hunk 1]
```diff
@@ -936,19 +847,25 @@ impl ServiceInner {
...
-        let addrs = self
-            .wait_for_peers(expectation, local_bootstrap_agent.our_address())
-            .await?;
-        info!(self.log, "Enough peers exist to enact RSS plan");
+        let maybe_sled_plan = SledPlan::load(&self.log).await?;
+        if let Some(plan) = &maybe_sled_plan {
+            let stored_peers: HashSet<Ipv6Addr> =
+                plan.sleds.keys().map(|a| *a.ip()).collect();
+            if stored_peers != bootstrap_addrs {
+                return Err(SetupServiceError::BadConfig("Set of sleds requested does not match those in existing sled plan".to_string()));
```
[the remark this answers is not in the document]
Comment (smklein) [reply sled-agent/src/rack_setup/service.rs:861]: Not that rustfmt seems to care about, apparently!

**Review by leftwo — APPROVED** [review]
Let's get this into main.  Dogfood wants it!

**Pull request merged**